rust/private/rust.bzl

```
@@ -1139,6 +1139,82 @@ rust_test_binary = rule(
         """),
 )
 
+def rust_test_suite(name, srcs, tags = [], **kwargs):
+    """A rule for creating a test suite for a set of `rust_test` targets.
+
+    This rule can be used for setting up typical rust [integration tests][it]. Given the following
+    directory structure:
+
+    ```text
+    [crate]/
+        BUILD.bazel
+        src/
+            lib.rs
+            main.rs
+        tests/
+            integrated_test_a.rs
+            integrated_test_b.rs
+            integrated_test_c.rs
+            patterns/
+                fibonacci_test.rs
+    ```
+
+    The rule can be used to generate [rust_test](#rust_test) targets for each source file under `tests`
+    and a [test_suite][ts] which encapsulates all tests.
+
+    ```python
+    load("//rust:defs.bzl", "rust_binary", "rust_library", "rust_test_suite")
+
+    rust_library(
+        name = "math_lib",
+        srcs = ["src/lib.rs"],
+    )
+
+    rust_binary(
+        name = "math_bin",
+        srcs = ["src/main.rs"],
+    )
+
+    rust_test_suite(
+        name = "integrated_tests_suite",
+        srcs = glob(["tests/**"]),
+        deps = [":math_lib"],
+    )
+    ```
+
+    [it]: https://doc.rust-lang.org/rust-by-example/testing/integration_testing.html
+    [ts]: https://docs.bazel.build/versions/master/be/general.html#test_suite
+
+    Args:
+        name (str): The name of the `test_suite`.
+        srcs (list): All test sources, typically `glob(["tests/**/*.rs"])`.
+        tags (list): Tags for both the test suite and the underlying [rust_test](#rust_test) target.
+        **kwargs (dict): Additional keyword arguments. Args from [rust_test](#rust_test) are parsed out and shared
+            between all tests while [common attributes](https://docs.bazel.build/versions/master/be/common-definitions.html#common-attributes)
+            are shared between the test targets and `test_suite`.
+    """
+    tests = []
+
+    for src in srcs:
+        if not src.endswith(".rs"):
+            fail("srcs should have `.rs` extensions")
+
+        # The test name should not end with `.rs` or have `/`.
+        test_name = src[:-3].replace("/", "_")
+        rust_test(
+            name = test_name,
+            srcs = [src],
+            tags = tags,
+            **kwargs
+        )
+        tests.append(test_name)
+
+    native.test_suite(
+        name = name,
+        tests = tests,
+        tags = tags,
```

Would you be alright going with:
```suggestion
        tags = kwargs.getattr("tags", None),
```
And then removing `tags` from the macro definition.  I've found that tends to reduce the overhead on the macro's definition in the long run as more arguments get shared, but need to have a default.

---

This seems to be opposite feedback from what @illicitonion gave. Am I misunderstanding this? Could either of you clarify for me?

---

If I'm interpreting correctly, I think it's slightly different.  In that case, it looks like the `tags` attribute was not being passed to `rust_test`.

What I'm suggesting here doesn't remove the tags from the tests, just changes how they're applied to the test_suit.

---

The suggestion didn't quite work but I believe I've addressed this.